topdown/crypto.go

```diff
@@ -378,42 +417,69 @@ func getX509CertsFromPem(pemBlocks []byte) ([]*x509.Certificate, error) {
 	return x509.ParseCertificates(decodedCerts)
 }
 
-func getRSAPrivateKeyFromString(key string) (interface{}, error) {
-	// if the input is PEM handle that
+func getPrivateKeyFromString(key string) (interface{}, error) {
 	if strings.HasPrefix(key, "-----BEGIN") {
-		return getRSAPrivateKeyFromPEM([]byte(key))
+		return getPrivateKeyFromPem(key)
 	}
 
-	// assume input is base64 if not PEM
 	b64, err := base64.StdEncoding.DecodeString(key)
 	if err != nil {
 		return nil, err
 	}
-
-	return getRSAPrivateKeyFromPEM(b64)
+	return getPrivateKeyFromPem(string(b64))
 }
 
-func getRSAPrivateKeyFromPEM(pemBlocks []byte) (interface{}, error) {
-
-	// decode the pem into the Block struct
-	p, _ := pem.Decode(pemBlocks)
-	if p == nil {
-		return nil, fmt.Errorf("failed to parse PEM block containing the key")
+func getPrivateKeyFromPem(pemBlocks string) (interface{}, error) {
+
+	var validPrivateKeys []crypto.PrivateKey
+
+	if len(pemBlocks) != 0 {
+
+		bs := []byte(pemBlocks)
+		for len(bs) > 0 {
+			inputLen := len(bs)
+			var block *pem.Block
+			block, bs = pem.Decode(bs)
+			if block == nil && len(bs) == 0 {
+				break
+			}
+
+			if inputLen == len(bs) {
+				break
+			}
+
+			if block == nil {
+				continue
+			}
+
+			switch block.Type {
+			case blockTypeRSAPrivateKey:
+
+				parsedKey, err := x509.ParsePKCS1PrivateKey(block.Bytes)
+				if err != nil {
+					return nil, err
+				}
+				validPrivateKeys = append(validPrivateKeys, parsedKey)
+			case blockTypePrivateKey:
+				parsedKey, err := x509.ParsePKCS8PrivateKey(block.Bytes)
+				if err != nil {
+					return nil, err
+				}
+				validPrivateKeys = append(validPrivateKeys, parsedKey)
+			case blockTypeEcPrivateKey:
+				parsedKey, err := x509.ParseECPrivateKey(block.Bytes)
+				if err != nil {
+					return nil, err
+				}
+				validPrivateKeys = append(validPrivateKeys, parsedKey)
+			}
+		}
 	}
-
-	// if the key is in PKCS1 format
-	if p.Type == blockTypeRSAPrivateKey {
-		return x509.ParsePKCS1PrivateKey(p.Bytes)
+	if len(validPrivateKeys) == 0 {
+		NoKeys := errors.New("No private keys found")
```

Comment:
I think in this case, we can just return an empty array.

Reply:
https://github.com/open-policy-agent/opa/pull/6084/commits/c5ca712950e2ca31a84487cd3ac120db94fd83d6